source/common/stats/thread_local_store.cc

```diff
@@ -566,9 +613,23 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogramFromStatNameWithTags(
                                      [&buckets, this](absl::string_view stat_name) {
                                        buckets = &parent_.histogram_settings_->buckets(stat_name);
                                      });
-    RefcountPtr<ParentHistogramImpl> stat(new ParentHistogramImpl(
-        final_stat_name, unit, parent_, *this, tag_helper.tagExtractedName(),
-        tag_helper.statNameTags(), *buckets));
+
+    RefcountPtr<ParentHistogramImpl> stat;
+    {
+      Thread::LockGuard lock(parent_.hist_mutex_);
+      auto iter = parent_.histogram_set_.find(final_stat_name);
+      if (iter != parent_.histogram_set_.end()) {
+        stat = RefcountPtr<ParentHistogramImpl>(*iter);
+      } else {
+        stat = new ParentHistogramImpl(final_stat_name, unit, parent_,
+                                       tag_helper.tagExtractedName(), tag_helper.statNameTags(),
+                                       *buckets, parent_.next_histogram_id_++);
+        if (!parent_.shutting_down_) {
```

Comment:
I suppose this is possible but did you hit this in a test out of curiosity? It would be good to cover this case if it matters.